### ISSUE TITLE
gradient error when tf.concat is used on a list of 0-D tensors (should not allow 0-D tensors concatenation at the first place)

### ISSUE THREAD

**Issue by lixilinx** [tensorflow/tensorflow#48762]
**System information**
Just a small corner case issue, not much related to OS, Python or hardware. 
- OS: Win10 and macOS 10.15.7
- TensorFlow version: 2.4.1
- Python version: 3.8.5

**Describe the current behavior**

When tf.concat is used to concat a list of 0-D tensors (although this very step should not be allowed), forward calculations have no issue, but reporting ZeroDivisionError when one tries to get the gradient.  

**Describe the expected behavior**

Either report error when tf.concat is used to concat a list of 0-D tensors as how numpy handles this situation (preferred), or fixed the gradient flow, or give more understandable error message (error message ZeroDivisionError seems somewhat misleading as /0 does not happen in the code).  

**Standalone code to reproduce the issue**
```
import tensorflow as tf

def this_works():
    x = tf.zeros([10])  
    with tf.GradientTape() as gt:
        gt.watch(x)
        a = []
        for i in range(10):
            a.append(x[i:i+1])
        y = tf.reduce_sum(tf.concat(a, axis=0))
    print(y)
    print(gt.gradient(y, x))

this_works()


def this_fails():
    x = tf.zeros([10])  
    with tf.GradientTape() as gt:
        gt.watch(x)
        a = []
        for i in range(10):
            a.append(x[i]) # should not allowed
        y = tf.reduce_sum(tf.concat(a, axis=0))
    print(y)
    print(gt.gradient(y, x)) # fails here, error message ZeroDivisionError 
    
this_fails()
```

**Comment by kumariko** [comment]
I was able to reproduce the issue in TF v2.6 .Please find the gist [**`here`**](https://colab.research.google.com/gist/kumariko/308d62717c40c0c4cedff8fcf1942d12/48762.ipynb#scrollTo=U1aCfIgn4nj3).
